File: api/server/src/main/java/su/plo/voice/api/server/audio/capture/BaseProximityServerActivation.java
package su.plo.voice.api.server.audio.capture;

import org.jetbrains.annotations.NotNull;
import su.plo.lib.api.server.permission.PermissionDefault;
import su.plo.voice.api.event.EventPriority;
import su.plo.voice.api.event.EventSubscribe;
import su.plo.voice.api.server.PlasmoVoiceServer;
import su.plo.voice.api.server.audio.line.ServerSourceLine;
import su.plo.voice.api.server.audio.source.ServerPlayerSource;
import su.plo.voice.api.server.event.audio.capture.ServerActivationRegisterEvent;
import su.plo.voice.api.server.event.audio.capture.ServerActivationUnregisterEvent;
import su.plo.voice.api.server.event.audio.source.ServerSourcePacketEvent;
import su.plo.voice.api.server.player.VoicePlayer;
import su.plo.voice.proto.data.audio.capture.VoiceActivation;
import su.plo.voice.proto.data.audio.line.VoiceSourceLine;
import su.plo.voice.proto.packets.tcp.clientbound.SourceAudioEndPacket;
import su.plo.voice.proto.packets.tcp.clientbound.SourceInfoPacket;
import su.plo.voice.proto.packets.tcp.serverbound.PlayerAudioEndPacket;
import su.plo.voice.proto.packets.udp.clientbound.SourceAudioPacket;
import su.plo.voice.proto.packets.udp.serverbound.PlayerAudioPacket;

import java.util.Optional;
import java.util.UUID;

public abstract class BaseProximityServerActivation {

    protected final PlasmoVoiceServer voiceServer;
    protected final String activationName;
    protected final UUID activationId;
    protected final UUID sourceLineId;
    protected final PermissionDefault defaultPermission;

    protected final SelfActivationInfo selfActivationInfo;

    public BaseProximityServerActivation(@NotNull PlasmoVoiceServer voiceServer,
                                         @NotNull String activationName,
                                         @NotNull PermissionDefault defaultPermission) {
        this.voiceServer = voiceServer;
        this.activationName = activationName;
        this.activationId = VoiceActivation.generateId(activationName);
        this.sourceLineId = VoiceSourceLine.generateId(activationName);
        this.defaultPermission = defaultPermission;
        this.selfActivationInfo = new SelfActivationInfo(voiceServer);
    }

    @EventSubscribe(priority = EventPriority.HIGHEST)
    public void onActivationRegister(@NotNull ServerActivationRegisterEvent event) {
        ServerActivation activation = event.getActivation();
        if (!activation.getName().equals(activationName)) return;

        voiceServer.getMinecraftServer()
                .getPermissionsManager()
                .register(getActivationPermission(), defaultPermission);
    }

    @EventSubscribe(priority = EventPriority.HIGHEST)
    public void onActivationUnregister(@NotNull ServerActivationUnregisterEvent event) {
        ServerActivation activation = event.getActivation();
        if (!activation.getName().equals(activationName)) return;

        voiceServer.getMinecraftServer()
                .getPermissionsManager()
                .unregister(getActivationPermission());
    }

    @EventSubscribe(priority = EventPriority.HIGHEST)
    public void onSourceSendPacket(@NotNull ServerSourcePacketEvent event) {
        if (!(event.getSource() instanceof ServerPlayerSource)) return;

        ServerPlayerSource source = (ServerPlayerSource) event.getSource();
        if (!selfActivationInfo.getLastPlayerActivationIds()
                .containsKey(source.getPlayer().getInstance().getUUID())
        ) {
            return;
        }

        if (event.getPacket() instanceof SourceInfoPacket) {
            selfActivationInfo.updateSelfSourceInfo(
                    source.getPlayer(),
                    source,
                    ((SourceInfoPacket) event.getPacket()).getSourceInfo()
            );
        } else if (event.getPacket() instanceof SourceAudioEndPacket) {
            source.getPlayer().sendPacket(event.getPacket());
        }
    }

    protected void sendAudioEndPacket(@NotNull ServerPlayerSource source,
                                      @NotNull PlayerAudioEndPacket packet) {
        sendAudioEndPacket(source, packet, packet.getDistance());
    }

    protected void sendAudioEndPacket(@NotNull ServerPlayerSource source,
                                      @NotNull PlayerAudioEndPacket packet,
                                      short distance) {
        SourceAudioEndPacket sourcePacket = new SourceAudioEndPacket(source.getId(), packet.getSequenceNumber());
        source.sendPacket(sourcePacket, distance);
    }

    protected void sendAudioPacket(@NotNull VoicePlayer player,
                                   @NotNull ServerPlayerSource source,
                                   @NotNull PlayerAudioPacket packet) {
        sendAudioPacket(player, source, packet, packet.getDistance());
    }

    protected void sendAudioPacket(@NotNull VoicePlayer player,
                                   @NotNull ServerPlayerSource source,
                                   @NotNull PlayerAudioPacket packet,
                                   short distance) {
        SourceAudioPacket sourcePacket = new SourceAudioPacket(
                packet.getSequenceNumber(),
                (byte) source.getState(),
                packet.getData(),
                source.getId(),
                distance
        );

        if (source.sendAudioPacket(sourcePacket, distance, packet.getActivationId())) {
            selfActivationInfo.sendAudioInfo(player, source, packet.getActivationId(), sourcePacket);
        }
    }

    protected Optional<ServerPlayerSource> getPlayerSource(@NotNull VoicePlayer player,
                                                         @NotNull UUID activationId,
                                                         boolean isStereo) {
        if (!activationId.equals(this.activationId)) return Optional.empty();

        Optional<ServerActivation> activation = voiceServer.getActivationManager()
                .getActivationById(activationId);
        if (!activation.isPresent()) return Optional.empty();

        Optional<ServerSourceLine> sourceLine = voiceServer.getSourceLineManager()
                .getLineById(sourceLineId);
        if (!sourceLine.isPresent()) return Optional.empty();

        isStereo = isStereo && activation.get().isStereoSupported();
        ServerPlayerSource source = voiceServer.getSourceManager().createPlayerSource(
                voiceServer,
                player,
                sourceLine.get(),
                "opus",
                isStereo
        );
        source.setLine(sourceLine.get());
        source.setStereo(isStereo);

        return Optional.of(source);
    }

    protected String getActivationPermission() {
        return "voice.activation." + activationName;
    }
}

File: api/server/src/main/kotlin/su/plo/voice/api/server/audio/capture/BaseProximityServerActivation.kt
package su.plo.voice.api.server.audio.capture

import com.google.common.collect.Maps
import su.plo.lib.api.server.permission.PermissionDefault
import su.plo.voice.api.event.EventPriority
import su.plo.voice.api.event.EventSubscribe
import su.plo.voice.api.server.PlasmoVoiceServer
import su.plo.voice.api.server.audio.source.ServerPlayerSource
import su.plo.voice.api.server.event.audio.capture.ServerActivationRegisterEvent
import su.plo.voice.api.server.event.audio.capture.ServerActivationUnregisterEvent
import su.plo.voice.api.server.event.audio.source.ServerSourcePacketEvent
import su.plo.voice.api.server.event.connection.UdpClientDisconnectedEvent
import su.plo.voice.api.server.event.player.PlayerActivationDistanceUpdateEvent
import su.plo.voice.api.server.player.VoiceServerPlayer
import su.plo.voice.proto.data.audio.capture.VoiceActivation
import su.plo.voice.proto.data.audio.line.VoiceSourceLine
import su.plo.voice.proto.packets.tcp.clientbound.SourceAudioEndPacket
import su.plo.voice.proto.packets.tcp.clientbound.SourceInfoPacket
import su.plo.voice.proto.packets.tcp.serverbound.PlayerAudioEndPacket
import su.plo.voice.proto.packets.udp.clientbound.SourceAudioPacket
import su.plo.voice.proto.packets.udp.serverbound.PlayerAudioPacket
import java.util.*
import java.util.function.Consumer

abstract class BaseProximityServerActivation(
    protected val voiceServer: PlasmoVoiceServer,
    protected val activationName: String,
    protected val defaultPermission: PermissionDefault
) {

    protected val activationId = VoiceActivation.generateId(activationName)
    protected val sourceLineId = VoiceSourceLine.generateId(activationName)
    protected val selfActivationInfo = SelfActivationInfo(voiceServer.udpConnectionManager)

    private val sourceByPlayerId: MutableMap<UUID, ServerPlayerSource> = Maps.newConcurrentMap()

    @EventSubscribe(priority = EventPriority.HIGHEST)
    fun onActivationRegister(event: ServerActivationRegisterEvent) {
        val activation = event.activation
        if (activation.name != activationName) return

        activation.permissions.forEach { permission ->
            voiceServer.minecraftServer
                .permissionsManager
                .register(permission, defaultPermission)
        }
    }

    @EventSubscribe(priority = EventPriority.HIGHEST)
    fun onActivationUnregister(event: ServerActivationUnregisterEvent) {
        val activation = event.activation
        if (activation.name != activationName) return

        activation.permissions.forEach { permission ->
            voiceServer.minecraftServer
                .permissionsManager
                .unregister(permission)
        }
    }

    @EventSubscribe(priority = EventPriority.HIGHEST)
    fun onSourceSendPacket(event: ServerSourcePacketEvent) {
        val source = event.source as? ServerPlayerSource ?: return

        if (!selfActivationInfo.lastPlayerActivationIds
                .containsKey(source.player.instance.uuid)
        ) {
            return
        }

        if (event.packet is SourceInfoPacket) {
            selfActivationInfo.updateSelfSourceInfo(
                source.player,
                source,
                (event.packet as SourceInfoPacket).sourceInfo
            )
        } else if (event.packet is SourceAudioEndPacket) {
            source.player.sendPacket(event.packet)
        }
    }

    @EventSubscribe
    fun onActivationDistanceChange(event: PlayerActivationDistanceUpdateEvent) {
        if (event.activation.id != VoiceActivation.PROXIMITY_ID) return
        if (event.oldDistance == -1) return
        event.player.visualizeDistance(event.distance)
    }

    @EventSubscribe
    fun onClientDisconnected(event: UdpClientDisconnectedEvent) =
        sourceByPlayerId.remove(event.connection.player.instance.uuid)

    protected fun sendAudioEndPacket(
        source: ServerPlayerSource,
        packet: PlayerAudioEndPacket,
        distance: Short = packet.distance
    ) {
        source.sendPacket(
            SourceAudioEndPacket(source.id, packet.sequenceNumber),
            distance
        )
    }

    protected fun sendAudioPacket(
        player: VoiceServerPlayer,
        source: ServerPlayerSource,
        packet: PlayerAudioPacket,
        distance: Short = packet.distance
    ) {
        val sourcePacket = SourceAudioPacket(
            packet.sequenceNumber, source.state.toByte(),
            packet.data,
            source.id,
            distance
        )

        if (source.sendAudioPacket(sourcePacket, distance, packet.activationId)) {
            selfActivationInfo.sendAudioInfo(player, source, packet.activationId, sourcePacket)
        }
    }

    protected fun getPlayerSource(
        player: VoiceServerPlayer,
        activationId: UUID,
        isStereo: Boolean?
    ): ServerPlayerSource? {
        if (activationId != this.activationId) return null

        val activation = voiceServer.activationManager
            .getActivationById(activationId)
            .orElse(null)
            ?: return null
        val sourceLine = voiceServer.sourceLineManager
            .getLineById(sourceLineId)
            .orElse(null)
            ?: return null

        return sourceByPlayerId.getOrPut(player.instance.uuid) {
            voiceServer.sourceManager.createPlayerSource(
                voiceServer,
                player,
                sourceLine,
                "opus",
                false
            )
        }.apply {
            line = sourceLine
        }.apply {
            isStereo?.let { isStereo ->
                setStereo(isStereo && activation.isStereoSupported)
            }
        }
    }
}
